### ISSUE TITLE
openstack kilo + ubuntu 14.4 cannot launch instances due to network failure to bind port

### ISSUE THREAD

**Issue by maguire-alan-e** [projectcalico/calico#1027]
Instance boot fails with an error in neutron server logs when trying to "nova boot" an instance:

2016-06-02 07:08:24.293 31171 INFO networking_calico.plugins.ml2.drivers.calico.t_etcd [req-a8ff1e4b-01cc-442d-a976-5ecaf66beb11 ] Deleting port {'status': u'DOWN', 'binding:host_id': u'controller', 'allowed_address_pairs': [], 'extra_dhcp_opts': [], 'device_owner': u'compute:None', 'binding:profile': {}, 'fixed_ips': [{'subnet_id': u'93dc9c4d-f5c0-43eb-8ba4-8e6582864382', 'ip_address': u'10.200.1.16'}], 'id': u'e740b622-c158-471c-a370-ce0d0486a7d4', 'security_groups': [u'6be669ea-7df0-4214-82fc-f7241498c597'], 'device_id': u'4b343ff1-d6cf-43f5-9efe-a1b173c1877b', 'name': u'', 'admin_state_up': True, 'network_id': u'67b582ee-5223-4fa5-99e3-97892357594c', 'tenant_id': u'f0061567548a42ccb70edc78437394c9', 'binding:vif_details': {}, 'binding:vnic_type': u'normal', 'binding:vif_type': u'binding_failed', 'mac_address': u'fa:16:3e:72:b2:49'}
2016-06-02 07:08:24.316 31171 ERROR neutron.plugins.ml2.managers [req-a8ff1e4b-01cc-442d-a976-5ecaf66beb11 ] Failed to bind port e740b622-c158-471c-a370-ce0d0486a7d4 on host compute-node

tar/gzipped diags attached. Let me know if I can provide any more info. Thanks!
[instance-boot-fail.tar.gz](https://github.com/projectcalico/calico/files/296039/instance-boot-fail.tar.gz)

**Comment by matthewdupre** [comment]
Looks like nova is having trouble connecting to neutron (from nova-compute.log.1):

```2016-06-02 01:26:59.623 7580 DEBUG oslo_concurrency.lockutils [-] Acquired semaphore "neutron_admin_auth_token_lock" lock /usr/lib/python2.7/dist-packages/oslo_concurrency/lockutils.py:391
2016-06-02 01:26:59.623 7580 DEBUG keystoneclient.auth.identity.v2 [-] Making authentication request to http://nws-bur-25-182:35357/v3/tokens get_auth_ref /usr/lib/python2.7/dist-packages/keystoneclient/auth/identity/v2.py:76
2016-06-02 01:26:59.648 7580 DEBUG keystoneclient.session [-] Request returned failure status: 404 request /usr/lib/python2.7/dist-packages/keystoneclient/session.py:388
2016-06-02 01:26:59.649 7580 DEBUG oslo_concurrency.lockutils [-] Releasing semaphore "neutron_admin_auth_token_lock" lock /usr/lib/python2.7/dist-packages/oslo_concurrency/lockutils.py:404
2016-06-02 01:26:59.649 7580 ERROR nova.compute.manager [-] Instance failed network setup after 1 attempt(s)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager Traceback (most recent call last):
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/nova/compute/manager.py", line 1795, in _allocate_network_async
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     dhcp_options=dhcp_options)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/nova/network/neutronv2/api.py", line 411, in allocate_for_instance
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     get_client(context, admin=True))
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/nova/network/neutronv2/api.py", line 205, in get_client
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     auth_token = _ADMIN_AUTH.get_token(_SESSION)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/auth/identity/base.py", line 104, in get_token
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     return self.get_access(session).auth_token
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/auth/identity/base.py", line 144, in get_access
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     self.auth_ref = self.get_auth_ref(session)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/auth/identity/v2.py", line 78, in get_auth_ref
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     authenticated=False, log=False)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/session.py", line 488, in post
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     return self.request(url, 'POST', **kwargs)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/utils.py", line 318, in inner
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     return func(*args, **kwargs)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager   File "/usr/lib/python2.7/dist-packages/keystoneclient/session.py", line 389, in request
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager     raise exceptions.from_response(resp, method, url)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager NotFound: The resource could not be found. (HTTP 404) (Request-ID: req-0e9860c3-8a02-42b1-bc74-0c15e9fa3c5f)
2016-06-02 01:26:59.649 7580 TRACE nova.compute.manager ```

**Comment by maguire-alan-e** [comment]
thanks for digging into this Matt! I should have mentioned this previously but the nova auth issues were a separate problem - I'm afraid I'm not having much luck. To fix that it turned out that nova needed to avoid the keystone API for authentication. Once that problem was fixed, nova boot got a bit further, and I encountered the network problem mentioned above. So it's reasonably likely this problem too is a nova/neutron misconfiguration on my part, but I'm not really sure where to start looking beyond the logs. The various pieces - etcd, felix etc seem to be working. One thing I did - and I'm not sure if this was correct - is to explicitly delete the neutron l3 agent, i.e. neutron agent-list looks like this now:

 neutron agent-list
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| id                                   | agent_type                    | host                         | alive | admin_state_up | binary                 |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| 4c44bfa5-984b-4b49-a7a8-1652376535e8 | Metadata agent                | nws-bur-25-182 | :-)   | True           | neutron-metadata-agent |
| 7206551d-0b32-4230-96b0-d92ba4120ac0 | DHCP agent                    | nws-bur25-182 | :-)   | True           | neutron-dhcp-agent     |
| c0d3c99e-3498-408f-a397-b5777f2558eb | Calico per-host agent (felix) | nws-bur-25-184               | :-)   | True           | calico-felix           |
| f8de15a5-b6a8-4a7e-9688-d5944b8e4e39 | Calico per-host agent (felix) | nws-bur-25-182               | :-)   | True           | calico-felix           |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+

**Comment by maguire-alan-e** [comment]
oops, meant nova needed to avoid the keystone v3 API above.

**Comment by matthewdupre** [comment]
You can indeed remove the l3 agent (also the metadata agent - we use nova-metadata-api / nova-api directly).

**Comment by matthewdupre** [comment]
One thing that might be helpful is a bit more information on the networks (etc) that are set up:

neutron net-list and net-show, plus subnet-list and subnet-show, router list.  A common way to hit the mysterious binding failed error is pre-Calico state hanging around.

**Comment by maguire-alan-e** [comment]
Thanks again for the quick response Matt! I only set up a single external network/subnet for initial sanity-checking; perhaps that could be part of the problem too? I had set up a router but removed it. Here's the output:

root@nws-bur-25-182:/var/log/neutron# neutron net-list
+--------------------------------------+----------+----------------------------------------------------+
| id                                   | name     | subnets                                            |
+--------------------------------------+----------+----------------------------------------------------+
| 67b582ee-5223-4fa5-99e3-97892357594c | external | 93dc9c4d-f5c0-43eb-8ba4-8e6582864382 10.200.1.0/24 |
+--------------------------------------+----------+----------------------------------------------------+
root@nws-bur-25-182:/var/log/neutron# neutron subnet-list
+--------------------------------------+-----------------+---------------+------------------------------------------------+
| id                                   | name            | cidr          | allocation_pools                               |
+--------------------------------------+-----------------+---------------+------------------------------------------------+
| 93dc9c4d-f5c0-43eb-8ba4-8e6582864382 | external-subnet | 10.200.1.0/24 | {"start": "10.200.1.2", "end": "10.200.1.254"} |
+--------------------------------------+-----------------+---------------+------------------------------------------------+
root@nws-bur-25-182:/var/log/neutron# neutron router-list

I tried removing it all and starting from scratch with an internal-only network, same pattern of
failure I'm afraid.

Is there perhaps more I should have done to create a basic sanity-checking topology? Thanks!

**Comment by matthewdupre** [comment]
Can I get net-show and subnet-show as well please?

**Comment by maguire-alan-e** [comment]
sure, here's the output for the internal-only net/subnet (which fails binding also):

root@nws-bur-25-182:/var/log/neutron# neutron net-show internal
+---------------------------+--------------------------------------+
| Field                     | Value                                |
+---------------------------+--------------------------------------+
| admin_state_up            | True                                 |
| id                        | 8bd61622-5fdf-4949-a443-7a8ba5baa01c |
| mtu                       | 0                                    |
| name                      | internal                             |
| provider:network_type     | local                                |
| provider:physical_network |                                      |
| provider:segmentation_id  |                                      |
| router:external           | False                                |
| shared                    | False                                |
| status                    | ACTIVE                               |
| subnets                   | 139d18e9-47f0-46af-96aa-9f90f6321668 |
| tenant_id                 | f0061567548a42ccb70edc78437394c9     |
+---------------------------+--------------------------------------+
root@nws-bur-25-182:/var/log/neutron# neutron subnet-show internal-subnet
+-------------------+----------------------------------------------------+
| Field             | Value                                              |
+-------------------+----------------------------------------------------+
| allocation_pools  | {"start": "192.168.20.2", "end": "192.168.20.254"} |
| cidr              | 192.168.20.0/24                                    |
| dns_nameservers   |                                                    |
| enable_dhcp       | True                                               |
| gateway_ip        | 192.168.20.1                                       |
| host_routes       |                                                    |
| id                | 139d18e9-47f0-46af-96aa-9f90f6321668               |
| ip_version        | 4                                                  |
| ipv6_address_mode |                                                    |
| ipv6_ra_mode      |                                                    |
| name              | internal-subnet                                    |
| network_id        | 8bd61622-5fdf-4949-a443-7a8ba5baa01c               |
| subnetpool_id     |                                                    |
| tenant_id         | f0061567548a42ccb70edc78437394c9                   |
+-------------------+----------------------------------------------------+
root@nws-bur-25-182:/var/log/neutron#

**Comment by matthewdupre** [comment]
Could I ask you to retry with neutron-server's log level turned up to DEBUG please?  There are a number of helpful logs in this area it might spit out.

**Comment by maguire-alan-e** [comment]
Sure. With debug neutron-server.log shows:

2016-06-02 22:29:02.872 11838 DEBUG neutron.plugins.ml2.managers [req-6b5dc218-bbb6-487f-a805-19dc098cf9c3 ] Attempting to bind port 79f8d74f-dfbe-456f-ab26-12a8ee84e011 on host nws-bur-25-184 for vnic_type normal with profile  bind_port /usr/lib/python2.7/dist-packages/neutron/plugins/ml2/managers.py:611
2016-06-02 22:29:02.873 11838 DEBUG neutron.plugins.ml2.managers [req-6b5dc218-bbb6-487f-a805-19dc098cf9c3 ] Attempting to bind port 79f8d74f-dfbe-456f-ab26-12a8ee84e011 on host nws-bur-25-184 at level 0 using segments [{'segmentation_id': None, 'physical_network': None, 'id': u'1b8c413d-a36b-4478-8dcd-043698239901', 'network_type': u'local'}] _bind_port_level /usr/lib/python2.7/dist-packages/neutron/plugins/ml2/managers.py:628
2016-06-02 22:29:02.874 11838 DEBUG neutron.plugins.ml2.drivers.mech_agent [req-6b5dc218-bbb6-487f-a805-19dc098cf9c3 ] Attempting to bind port 79f8d74f-dfbe-456f-ab26-12a8ee84e011 on network 8bd61622-5fdf-4949-a443-7a8ba5baa01c bind_port /usr/lib/python2.7/dist-packages/neutron/plugins/ml2/drivers/mech_agent.py:60
2016-06-02 22:29:02.879 11838 ERROR neutron.plugins.ml2.managers [req-6b5dc218-bbb6-487f-a805-19dc098cf9c3 ] Failed to bind port 79f8d74f-dfbe-456f-ab26-12a8ee84e011 on host nws-bur-25-184
2016-06-02 22:29:02.880 11838 ERROR neutron.plugins.ml2.managers [req-6b5dc218-bbb6-487f-a805-19dc098cf9c3 ] Failed to bind port 79f8d74f-dfbe-456f-ab26-12a8ee84e011 on host nws-bur-25-184

**Comment by matthewdupre** [comment]
OK, looks like neutron/ml2/drivers/mech_agent.py's bind_port isn't finding any host_agents (we see the Attempting to bind, but no Checking agent).

Can you do a `neutron agent-list` please?

**Comment by maguire-alan-e** [comment]
sure (apologies for the delay, I'm based in Ireland): 

root@nws-bur-25-182:/var/log/neutron# neutron agent-list
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| id                                   | agent_type                    | host                         | alive | admin_state_up | binary                 |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| 2298e65e-cbcb-4f54-bbf8-a58502b50006 | Metadata agent                | nws-bur-25-184 | :-)   | True           | neutron-metadata-agent |
| 455d75d8-a577-4438-88ad-6cc6a1ec9db1 | DHCP agent                    | nws-bur-25-184 | :-)   | True           | neutron-dhcp-agent     |
| 4c44bfa5-984b-4b49-a7a8-1652376535e8 | Metadata agent                | nws-bur-25-182 | :-)   | True           | neutron-metadata-agent |
| 7206551d-0b32-4230-96b0-d92ba4120ac0 | DHCP agent                    | nws-bur-25-182 | :-)   | True           | neutron-dhcp-agent     |
| c0d3c99e-3498-408f-a397-b5777f2558eb | Calico per-host agent (felix) | nws-bur-25-184               | :-)   | True           | calico-felix           |
| f8de15a5-b6a8-4a7e-9688-d5944b8e4e39 | Calico per-host agent (felix) | nws-bur-25-182               | :-)   | True           | calico-felix           |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+

I noticed there's some complaints in felix-etcd log on the compute node which the instance is being spawned on ("bad HTTP status topol"l, there's a lot of resyncing going on too - not sure if that's normal though.

2016-06-03 08:50:41,929 [INFO][8114/140593928926976] urllib3.connectionpool 188: Starting new HTTP connection (4): localhost
2016-06-03 08:50:41,932 [WARNING][8114/140593928926976] calico.etcddriver.driver 782: etcd watch returned bad HTTP status topoll on index 37843: 400
2016-06-03 08:50:41,933 [ERROR][8114/140593928926976] calico.etcddriver.driver 821: Error from etcd for index 37843: {'errorCode': 401, 'index': 38862, 'message': 'The event in requested index is outdated and cleared', 'cause': 'the requested history has been cleared [37863/37843]'}; triggering a resync.
2016-06-03 08:50:41,933 [INFO][8114/140593928926976] calico.etcddriver.driver 885: STAT: Final watcher etcd response time: 0 in 3513.0s (0.000/s) min=0.000ms mean=0.000ms max=0.000ms
2016-06-03 08:50:41,934 [INFO][8114/140593928926976] calico.etcddriver.driver 885: STAT: Final watcher processing time: 9 in 3513.0s (0.003/s) min=0.662ms mean=0.753ms max=1.006ms
2016-06-03 08:50:41,934 [INFO][8114/140593928926976] calico.etcddriver.driver 888: Watcher thread finished. Signalled to resync thread. Was at index 37843.  Queue length is 1.
2016-06-03 08:50:41,961 [WARNING][8114/140593939724032] calico.etcddriver.driver 279: Watcher died; resyncing.

**Comment by fasaxc** [comment]
@maguire-alan-e the bonus resyncs are a known issue but they should be harmless; they happen when the system is idle for a long time.  (The issue is that we watch etcd for changes to the `/calico/v1` directory; if there are no changes inside that directory for a long time but there are enough changes outside then etcd's event buffer wraps.)

**Comment by maguire-alan-e** [comment]
@fasaxc thanks for the explanation!

**Comment by maguire-alan-e** [comment]
one quick update - I noticed that creation of the virtual disk during instance creation failed in a way matching the following issue

https://bugs.launchpad.net/openstack-ansible/+bug/1428639  Nova requires python-libguestfs in Kilo 

I've added the packages.required on control/compute nodes, ran " update-guestfs-appliance" 
and then encountered

https://bugs.launchpad.net/devstack/+bug/1413142  bad configuration for libguestfs 

With the permissions fix and nova-compute restarts, the problems with disk creation go away.
However the "Unexpected vif_type" failure as originally reported persists I'm afraid.

**Comment by matthewdupre** [comment]
Sorry to leave you hanging on this one - it's not obvious to me where it's going wrong.

It's horrible I know, but it might be easiest to add some extra logging to neutron/ml2/drivers/mech_agent.py - it'll be in /usr/lib/python2.7/site-packages/ or similar.  I'd be particularly interested in what's happening in bind_port() - it seems like context.host_agents() must be empty, but I'm not sure how that is - it seems like you have the Felixes reporting themselves as alive.

**Comment by maguire-alan-e** [comment]
Thanks Matt - I've added a bit of debugging and it appears that in bind_port() the vnic_type
is "normal", self.agent_type is reported as "Calico per-host agent (felix)", but there appear to be
no .context.host_agents for this agent type, which explains why we don't see the "Checking
agent" message. Here's the debugging I added:


        LOG.debug("Got vnic_type of %s", vnic_type)
        LOG.debug("Got agent type %s", self.agent_type)
        for agent in context.host_agents(self.agent_type):
            LOG.debug("Got agent %s", agent)

We see output for the vnic_type, agent_type but nothing from the debug logging in the loop.

"neutron agent-list" reports the agents as "Calico per-host agent (felix)", and they appear to be
up:

# neutron agent-list
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| id                                   | agent_type                    | host                         | alive | admin_state_up | binary                 |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
| 0e5bc24b-1eab-4bae-9485-a6e99cb8b1d0 | Metadata agent                | nws-bur-25-182 | :-)   | True           | neutron-metadata-agent |
| 15e5318f-6164-42ea-b622-48ea7154eb74 | Calico per-host agent (felix) | nws-bur-25-116               | :-)   | True           | calico-felix           |
| 171004fe-6cfd-48d9-a43d-42d95c8bf42b | DHCP agent                    | nws-bur-25-182 | :-)   | True           | neutron-dhcp-agent     |
| a1499ef3-ce4c-47f4-8053-dae0365b1cc6 | Metadata agent                | nws-bur-25-184 | :-)   | True           | neutron-metadata-agent |
| a463192e-c53a-44ad-b67d-ce5ae23c4a8b | L3 agent                      | nws-bur-25-182 | :-)   | True           | neutron-l3-agent       |
| c0d3c99e-3498-408f-a397-b5777f2558eb | Calico per-host agent (felix) | nws-bur-25-184               | :-)   | True           | calico-felix           |
| c5230db7-806f-436a-8caa-fcd1d9e27b01 | DHCP agent                    | nws-bur-25-116 | :-)   | True           | neutron-dhcp-agent     |
| c55eab1c-53ed-41a4-b3b6-fb6d11c11973 | DHCP agent                    | nws-bur-25-184 | :-)   | True           | neutron-dhcp-agent     |
| f8de15a5-b6a8-4a7e-9688-d5944b8e4e39 | Calico per-host agent (felix) | nws-bur-25-182               | :-)   | True           | calico-felix           |
+--------------------------------------+-------------------------------+------------------------------+-------+----------------+------------------------+
root@nws-bur-25-182:/usr/lib/python2.7/dist-packages/neutron/plugins/ml2/drivers# 

So I guess the question is why the context.host_agents is empty? I'll try and figure out where
to add additional debugging to chase this up; let me know if you have any additional suggestions
and thanks again for the help!

**Comment by maguire-alan-e** [comment]
I've got things working at last. I can't say I'm 100% sure that this is what solved the problem (as VM boot did not work immediately after I made this change), but changing the /etc/calico/felix.cfg to a
fully-qualified domain name on each compute node seemed to fix things. I can boot a VM
now and it has a DHCP-acquired address from the internal network. I'm guessing that the filtering 
of agents failed on the basis of not matching a FQDN with an unqualified hostname - if this is the root
cause would  it make sense to mention this potential issue in the docs? This issue could be repurposed 
as a doc bug if so I guess. Thanks again for all the help!
